logstash-core/lib/logstash/environment.rb

```diff
@@ -1,7 +1,40 @@
 # encoding: utf-8
 require "logstash/errors"
+require "logstash/config/cpu_core_strategy"
+require "logstash/settings"
 
 module LogStash
+
+  [
+            Setting::String.new("node.name", Socket.gethostname),
+           Setting::Boolean.new("config.allow_env", false),
+            Setting::String.new("config.path", nil, false),
+            Setting::String.new("config.string", nil, false),
+           Setting::Boolean.new("config.test", false),
+           Setting::Boolean.new("config.reload.auto", false),
+           Setting::Numeric.new("config.reload.interval", 3),
+           Setting::Boolean.new("metric.collect", true) {|v| v == true }, # metric collection cannot be disabled
+            Setting::String.new("settings.dir", ::File.join(Environment::LOGSTASH_HOME, "conf")),
+            Setting::String.new("pipeline.id", "main"),
+           Setting::Numeric.new("pipeline.workers", LogStash::Config::CpuCoreStrategy.maximum),
+           Setting::Numeric.new("pipeline.output.workers", 1),
+           Setting::Numeric.new("pipeline.batch.size", 125),
+           Setting::Numeric.new("pipeline.batch.delay", 5), # in milliseconds
+           Setting::Boolean.new("pipeline.unsafe_shutdown", false),
+                    Setting.new("plugin.paths", Array, []),
+            Setting::String.new("interactive", nil, false),
+           Setting::Boolean.new("debug", false),
+           Setting::Boolean.new("debug.config", false),
+           Setting::Boolean.new("verbose", false),
+           Setting::Boolean.new("quiet", false),
+           Setting::Boolean.new("version", false),
+           Setting::Boolean.new("help", false),
+            Setting::String.new("log.path", nil, false),
+           Setting::Boolean.new("log.json", false),
+            Setting::String.new("web_api.http.host", "127.0.0.1"),
+              Setting::Port.new("web_api.http.port", 9600),
+  ].each {|setting| SETTINGS.register(setting) }
```

Comment:
I don't think this list should live in settings.rb since that file just creates the concepts for settings handling. This block of text actually defines settings important for LogStash, before they're needed in the runner, which is the environment.

As for inferring, I'm not sure I understand. in a setting like `Setting::Boolean.new("debug", false)`, you want to just do `Setting.new("debug", false)` and infer it's a Setting::Boolean from the default value's class? I think we could do it, but it would be inconsistent with the settings that have no default value and we cannot infer those.